### ISSUE TITLE
Uncaught Error: com.facebook.react.devsupport.JSException: Unexpected end of script

### ISSUE THREAD

**Issue by AdamJNavarro** [expo/expo#3674]
Related forum post: https://forums.expo.io/t/attempting-to-call-js-function-on-a-bad-application-bundle/20142

Possible related issue: https://github.com/expo/expo/issues/1987

**Comment by somangshu** [comment]
Any updates on this @AdamJNavarro. I am having this issue on Expo v32, Expokit 32.0.7. As usual this issue can be found only on some particular devices. 

Kindly point to any possible workarounds you may have come accross

**Comment by randomjunk** [comment]
TLDR; I think this is an intermittent issue where the bundle download is corrupted. Resetting app data fixes a broken app instance.

Repost of my forum comment:

The original error (Unexpected end of script) also happened for at least one of our customers after the last OTA update we made (SDK 32 with a binary built in the last few days).

They were an Android user and clearing the app data and cache solved the issue (from server logs the app must have re-downloaded the OTA update correctly when it was restarted).

So this appears to have been an issue where the OTA update download is interrupted leaving a corrupt file – the Expo runtime seems not to be able to detect this or recover from it gracefully.
Looking at the code (I think here? Exponent.java) if the app is killed at that line we'll potentially end up with a partially written bundle file which the runtime cannot recover from unless the bundle is updated again.

A fix for this would probably be to re-trigger bundle download if runApplication throws the JSException. And/or write a bundle checksum first and verify that before trying to use it.

This certainly isn’t a common problem, but with thousands of users, it is turning up fairly frequently. We’re asking them to clear app data and re-login which gets them going again but obviously doesn’t look very good.

**Comment by brentvatne** [comment]
cc @esamelson as you worked on ota updates

**Comment by esamelson** [comment]
thanks all, sorry for the slow followup. i'll look into this as soon as i can!

**Comment by esamelson** [comment]
Hey all. First off I want to apologize for this issue -- we understand it's frustrating to run into issues like this with something as simple as app loading. We know the limitations of the current system, and we're planning to do a significant revamp and rewrite of the OTA updates/code loading system in the clients and standalone apps to focus more on atomicity, consistency, and proper handling of issues like this across the board. However, this is likely on the timescale of 6 months- end of the year and will not be helpful in addressing this particular issue right now.

With regards to this, we do have an error handling system in place for OTA updates. The expected behavior is that if a given OTA update errors on its first run, the app will mark it as unsafe and revert back to the previous version of the app. It sounds like that mechanism may not working correctly in this particular case. If this issue is important to you, I'd recommend doing some more digging and trying to find the root cause of this issue -- i.e. why the error handling system is failing in this case.

Hope this helps, and sorry not to have a better answer!

**Comment by cruzach** [comment]
I'm going to close this for now since we have replaced the legacy Updates module with `expo-updates`- but if you see this same behavior with the new module please open a new issue
